src/generic/complex_matrices.cc

```diff
@@ -538,6 +539,647 @@ namespace oomph
     }
   }
 
+  //===========================================================================
+  /// Function to multiply this matrix by the CRDoubleMatrix matrix_in.
+  /// In a serial matrix, there are 4 methods available:
+  /// Method 1: First runs through this matrix and matrix_in to find the storage
+  ///           requirements for result - arrays of the correct size are
+  ///           then allocated before performing the calculation.
+  ///           Minimises memory requirements but more costly.
+  /// Method 2: Grows storage for values and column indices of result 'on the
+  ///           fly' using an array of maps. Faster but more memory
+  ///           intensive.
+  /// Method 3: Grows storage for values and column indices of result 'on the
+  ///           fly' using a vector of vectors. Not particularly impressive
+  ///           on the platforms we tried...
+  /// Method 2 is employed by default.
+  //=============================================================================
+  void CRComplexMatrix::multiply(const CRComplexMatrix& matrix_in,
+                                 CRComplexMatrix& result) const
+  {
+#ifdef PARANOID
+    // check that this matrix is built
+    if (!Built)
+    {
+      std::ostringstream error_message_stream;
+      error_message_stream << "This matrix has not been built";
+      throw OomphLibError(error_message_stream.str(),
+                          OOMPH_CURRENT_FUNCTION,
+                          OOMPH_EXCEPTION_LOCATION);
+    }
+    // check that this matrix is built
+    if (!matrix_in.built())
+    {
+      std::ostringstream error_message_stream;
+      error_message_stream << "This matrix matrix_in has not been built";
+      throw OomphLibError(error_message_stream.str(),
+                          OOMPH_CURRENT_FUNCTION,
+                          OOMPH_EXCEPTION_LOCATION);
+    }
+    // if soln is setup then it should have the same distribution as x
+    if (result.built())
+    {
+      if (!(*result.distribution_pt() == *this->distribution_pt()))
+      {
+        std::ostringstream error_message_stream;
+        error_message_stream
+          << "The matrix result is setup and therefore must have the same "
+          << "distribution as the vector x";
+        throw OomphLibError(error_message_stream.str(),
+                            OOMPH_CURRENT_FUNCTION,
+                            OOMPH_EXCEPTION_LOCATION);
+      }
+    }
+#endif
+
+    // short name for Serial_matrix_matrix_multiply_method
+    unsigned method = 2;
+
+    // if this matrix is not distributed and matrix in is not distributed
+    if ((method == 1) || (method == 2) || (method == 3))
+    {
+      // NB N is number of rows!
+      unsigned long N = this->nrow();
+      unsigned long M = matrix_in.ncol();
+      unsigned long Nnz = 0;
+
+      // pointers to arrays which store result
+      int* Row_start = 0;
+      std::complex<double>* Value = 0;
+      int* Column_index = 0;
```

Comment:
These variables aren't private variables of the current matrix, so should they be made lowercase?

Reply:
Yeah. I'm not sure what is going on here. It seems we are overriding private variables of the current matrix with local function variables...
Value, N, M, Nnz are defined in the SparseMatrix class
Row_start, Column_index in the CRMatrix class which inherits SparseMatrix
CRComplexMatrix then inherits CRMatrix

Reply:
Not overriding; redefining in the local scope. Possibly a copy and paste job from a constructor or build function which actually sets the internal private variable. Should all be okay if you make them lowercase (and propagate the case changes through the code)

**Example:**
```cpp
#include <iostream>
#include <string>

class A
{
public:
    A(const int& v=0) : V(v) {}
    void do_something(const int& new_v)
    {
        int V = new_v;
        printf("do_something(): V=%d\n", V);
    }
    void print_v()
    {
        printf("print_v(): V=%d\n", V);
    }
private:
    int V;
};

int main()
{
    A a;
    a.print_v();
    a.do_something(100);
    a.print_v();
}

# print_v(): V=0
# do_something(): V=100
# print_v(): V=0

```

Reply:
Fixed